---
title: "SCLC Analysis"
author: "Andrew Willems & Tian Hong"
date: "2023-04-05"
output: html_document
---

```{r setup, include=FALSE}
# This sets the global chunk options for knitr to display the R code.
knitr::opts_chunk$set(echo = TRUE)

# This sets the root directory for knitr to the project directory.
knitr::opts_knit$set(root.dir = "~/Documents/Work/Phd_program/hong_lab/Projects/sclc/")
```

## SCLC Analysis

```{r load needed packages}
# Load necessary packages using the pacman package
# data.table: for efficient data manipulation
# devtools: for developing R packages
# ggplot2: for data visualization
# glmnet: for fitting generalized linear models via penalized maximum likelihood
# hoardr: for caching data to disk
# Matrix: for sparse matrix operations
# monocle3: for single-cell analysis and visualization
# phateR: for visualizing high-dimensional data
# reticulate: for running Python code in R
# retry: for allowing us to repeatedly restart code due to timeouts
# Rmagic: for running R code in Jupyter notebooks
# styler: for consistent code formatting
# survminer: for survival analysis and visualization
# tidyverse: for data manipulation and visualization
# viridis: for creating color palettes for data visualization

pacman::p_load(
  cowplot, data.table, devtools, ggplot2, glmnet, hoardeR, igraph, Matrix, monocle3, parallel, phateR,
  reticulate, retry, Rmagic, styler, survminer, svglite, switchde, tidyverse, viridis
)
```

```{r miniconda setup for MAGIC}
# Loading conda environment 'r-reticulate' for MAGIC package
use_condaenv("/Users/andrewwillems/Library/r-miniconda-arm64/envs/r-reticulate")
```


```{r sourcing needed functions}
# This code sources several R scripts that contain functions used for analyzing miRNA expression data in lung cancer.
# The 'back_elim.R' script contains a function for performing backwards variable selection on the Cox model to identify the most significant miRNAs associated with survival.
# The 'cox_model.R' script contains a function for performing a Cox proportional hazards model on the data.
# The 'km_plotter.R' script contains a function for creating a Kaplan-Meier plot to visualize survival data.
# The 'lung_mirna_calculator.R' script contains functions for preprocessing and normalizing the miRNA expression data.
# The 'risk_score_calculator.R' script contains a function for calculating a risk score for each patient based on the results of the Cox model.
# This code also sources 'cell_dataset_builder.R', which contains functions for building a dataset from processed scRNA-seq expression data.

invisible(lapply(list.files("Code", pattern = "\\.R$", full.names = TRUE), source))
```


```{r reading data}
# Load gene expression data from a file and select relevant columns
scl_df <- read.csv(
  file = "Data/sclc_ucologne_2015/data_mrna_seq_rpkm.txt",
  sep = "\t"
)
scl_df <- subset(scl_df,
  select = c(Hugo_Symbol, sclc_ucologne_2015_S00022:sclc_ucologne_2015_S02397)
)

# Clean up column names by removing prefix
colnames(scl_df) <- sub("^sclc_ucologne_2015_S", "", colnames(scl_df))

# Load clinical data for patients and clean up column names and values
scl_patient <- read.csv(
  file = "Data/sclc_ucologne_2015/data_clinical_patient.txt",
  sep = "\t"
)

scl_patient <- scl_patient %>%
  slice(5:121) %>%
  rename(
    Patient.ID = X.Patient.Identifier, Overall.Survival.Status = Overall.Survival.Status,
    Overall.Survival.Months. = Overall.Survival..Months., Smoker = Smoker
  ) %>%
  select(Patient.ID, Sex, Overall.Survival.Status, Overall.Survival.Months., Smoker) %>%
  filter(Smoker != "Never") %>%
  filter(Smoker != "") %>%
  mutate(
    Smoker = as.factor(Smoker),
    Smoker = if_else(Smoker == "Current", true = 1, false = 0)
  ) %>%
  mutate(Overall.Survival.Status = ifelse(Overall.Survival.Status == "1:DECEASED", 1, 0)) %>%
  rename(
    vital.status = Overall.Survival.Status,
    time = Overall.Survival.Months.
  ) %>%
  mutate(Patient.ID = gsub(
    x = Patient.ID,
    pattern = "sclc_ucologne_2015_S",
    replacement = ""
  )) %>%
  mutate(vital.status_new = as.numeric(vital.status)) %>%
  mutate(time_new = as.numeric(time)) %>%
  select(-vital.status, -time) %>%
  rename(
    vital.status = vital.status_new,
    time = time_new
  ) %>%
  mutate(time = time * 30) %>%
  filter(time != 0) %>%
  rename_all(tolower)

# Extract the list of gene symbols
scl_genes <- scl_df$Hugo_Symbol

# Find the list of patients that have data for both gene expression and clinical data
common_patients <- intersect(scl_patient$patient.id, colnames(scl_df))

# Select only the common columns from gene expression data and transpose it
# so that each row corresponds to a patient
scl_common <- scl_df %>%
  select(all_of(common_patients)) %>%
  t() %>%
  as.data.frame()

# Assign the gene symbols as column names
colnames(scl_common) <- scl_genes

# Remove duplicated columns and add patient ID as a column
scl_common <- scl_common %>%
  select(-which(duplicated(names(.)))) %>%
  rownames_to_column(var = "patient.id")

# Join the gene expression data with the clinical data for each patient
scl_common <- left_join(scl_common, scl_patient[, c(
  "sex", "smoker",
  "vital.status", "time",
  "patient.id"
)],
by = "patient.id"
) %>%
  rename_all(tolower)

# Removing intermediate files to keep the environment clean
rm(scl_df, scl_patient, scl_genes, common_patients)
```


```{r single-cell data loading}
# Read in the gene expression data and filter out columns that start with "SCLC_P1" or "SCLC_P14"
sc_df <- fread("Data/OMIX002441-01.csv")
sc_df <- sc_df %>%
  select(-matches("^(SCLC_P1|SCLC_P14)")) %>%
  t() # transpose the data frame to have cells as rows

# Set column names to the first row of the data frame
colnames(sc_df) <- sc_df["V1", ]
sc_df <- sc_df[-1, ] # remove first row, which is now redundant
sc_df <- as.data.frame(sc_df)

# Read in the metadata and filter for tumor samples only
meta_data <- fread("Data/OMIX002441-02.csv")
meta_data <- meta_data %>%
  filter(NT == "tumor")

# Set row names to the first column, which contains the cell IDs
colnames(meta_data)[1] <- "id"

# Join the two dataframes on the "id" column
sc_df$id <- rownames(sc_df)
sc_df <- inner_join(sc_df, meta_data, by = "id")
rownames(sc_df) <- sc_df$id
sc_df <- sc_df %>% select(-id, -patient, -NT, -tissue, -cell_type, -nCount_RNA, -nFeature_RNA, -Pre_Map_Reads, -Aligned_Reads, -MappingRate)

# Keep only genes that are expressed in at least 10 cells
keep_cols <- colSums(sc_df > 0) > 10
sc_df <- sc_df[, keep_cols]

sc_rownames <- rownames(sc_df)

# Convert the data frame to integers
sc_df <- apply(sc_df, c(1, 2), as.integer) %>%
  as.data.frame()

rownames(sc_df) <- sc_rownames

# Plot the distribution of library sizes
p1 <- ggplot() +
  geom_histogram(aes(x = rowSums(sc_df)), bins = 50) +
  geom_vline(xintercept = 1000, color = "red") +
  theme(
    panel.background = element_blank(),
    axis.title = element_text(size = 22, face = "bold"),
    axis.text = element_text(size = 20),
    plot.title = element_text(size = 24, face = "bold", hjust = 0.5),
    axis.ticks = element_line(linewidth = 1.25),
    axis.ticks.length = unit(0.2, units = "cm"),
    axis.line = element_line(color = "black", linewidth = 1.25),
    plot.subtitle = element_text(hjust = 0.5, size = 22)
  ) +
  ggtitle("Distribution of Library Sizes in scRNA-seq", subtitle = "Distribution of library sizes after keeping genes that have positive (> 0) expression in at least 10 cells") +
  xlab("Library Size") +
  ylab("Count") +
  scale_y_continuous(expand = c(0, 0)) +
  scale_x_continuous(expand = c(0, 0))

ggsave(plot = p1, width = 20, height = 20, units = "in", dpi = 600, device = "png", filename = "sc_library_distribution_sizes.png", path = "Outputs/")
ggsave(plot = p1, width = 20, height = 20, units = "in", dpi = 600, device = "svg", filename = "sc_library_distribution_sizes.svg", path = "Outputs/")

# Keep only cells with a library size between 1000 and 15000
keep_rows <- rowSums(sc_df) > 1000 & rowSums(sc_df) < 15000
sc_df <- sc_df[keep_rows, ]

# Normalize the library size and take the square root of the expression values
sc_df <- library.size.normalize(sc_df)
sc_df <- sqrt(sc_df)

# Removing intermediate files to keep the environment tidy
rm(keep_cols, keep_rows, sc_rownames, p1, scl_common)
```

```{r single-cell denoising}
# Convert the sc_df dataframe to a matrix
sc_df_mat <- as.matrix(sc_df)

# Apply the MAGIC (Model-based Analysis of Genome-wide CRISPR/Cas9 Knockout) algorithm to denoise the scRNA-seq data.
# The 'genes' parameter specifies the set of genes to use for MAGIC imputation. Here, 'all_genes' is used to include all genes in the analysis.
# The 'seed' parameter specifies a random seed for reproducibility.
# The 'n.jobs' parameter specifies the number of cores to use for parallelization. Setting it to -1 indicates that all available cores should be used.
sc_denoised <- magic(sc_df_mat, genes = "all_genes", seed = 123, n.jobs = -1)

saveRDS(sc_denoised, "Outputs/denoised_sc_data.rds")

rm(sc_df_mat)
```


```{r generating pseudotime from single-cell data}
sc_denoised$result <- sc_denoised$result %>%
  t()

meta_data <- meta_data %>%
  filter(id %in% colnames(sc_denoised$result)) %>%
  column_to_rownames(var = "id")

cds_cell_meta <- meta_data
cds_gene_meta <- data.frame(gene_short_name = rownames(sc_denoised$result))
rownames(cds_gene_meta) <- cds_gene_meta$gene_short_name

emt_genes <- read.csv("Data/emt_genes_tian.txt", sep = "\t")

# Getting common genes between EMT list and our list
common_emt_genes <- intersect(cds_gene_meta$gene_short_name, emt_genes$Gene)
common_emt_genes <- data.frame(gene_short_name = common_emt_genes)
rownames(common_emt_genes) <- common_emt_genes$gene_short_name

cds_output <- cell_dataset_builder(
  gene_expression = "TUBA1A",
  cell_data = sc_denoised$result,
  gene_meta = cds_gene_meta,
  cell_meta = cds_cell_meta,
  pt_root = "Y_7",
  gene_title = "TUBA1A",
  pt_title = "TUBA1A",
  point_size = 2.5,
  norm_flag = "none",
  use_ordering_funct = FALSE,
  cds_filename = "Outputs/cds.rds",
  plot_file = "Outputs/emt_gene_expression.png",
  pt_file = "Outputs/pt_expression.png",
  pt_data_filename = "Outputs/TUBA1A_pseudotime_data.csv"
)

# Bringing the cell expression and pseudotime graphs together into a single plot
combo_plot <- ggarrange(cds_output$Cell_Progression_Graph, cds_output$PT_Graph, ncol = 2, nrow = 1, align = "h", labels = c("A.", "B."), legend = "bottom", font.label = c(size = 18))
combo_plot <- combo_plot + ggtitle("SDE Pseudotime Ordering") +
  theme(
    plot.title = element_text(size = 22, face = "bold", hjust = 0.5),
    plot.subtitle = element_text(hjust = 0.5, size = 9)
  )

# add a caption using ggdraw
caption_text <- "Figure S1: Panel A shows the gene expression levels of TUBA1A, which is a marker for mesenchymal cells during epithelial-mesenchymal transition (EMT). This gene's expression levels were used as the basis for calculating pseudotime in our single-cell RNA sequencing (scRNA-seq) dataset.

Panel B shows the calculated pseudotime of the scRNA-seq dataset, based on the expression levels of TUBA1A in panel A. The annotations of 1 and 2 (seen in white circles) indicate the roots of the pseudotime calculations, where the pseudotime is equal to 0. As we move along the trajectory (represented by the solid black line), the pseudotime values increase, mirroring the expression pattern of TUBA1A. "
caption <- ggdraw() +
  draw_label(caption_text, fontface = "plain", size = 10, x = 0, hjust = 0)

# add the caption to the bottom of the plot
plots_with_caption <- plot_grid(combo_plot, caption, nrow = 2, align = "v", axis = "tblr")


ggsave(
  plot = combo_plot, device = "png", dpi = 600, width = 7, height = 5,
  units = "in", filename = "Outputs/sde_foundation_plot.png", bg = "white"
)

ggsave(
  plot = combo_plot, device = "svg", dpi = 600, width = 7, height = 5,
  units = "in", filename = "Outputs/sde_foundation_plot.svg", bg = "white"
)
```

```{r sde metric}
sde_genes <- switchde_calculator(sc_denoised$result,
  pseudo_time = cds_output$PT_Data
)

saveRDS(sde_genes, "Outputs/sde_genes.rds")

head(sde_genes)
```


```{r mad metric}
mad_genes <- mad_calculator(sc_denoised$result)
saveRDS(mad_genes, "Outputs/mad_genes.rds")
head(mad_genes)
```



```{r mirna}
# Load the data
mirna_data <- read.csv("Data/mirna_data/hmdd_3_2_all_data.txt", sep = "\t")

# Filter the data down to just miRNAs that are lung related and are
# classified as upregulated/increased
mirna_data <- mirna_data %>%
  filter(grepl("up", category)) %>%
  filter(disease %in% c(
    "Lung Disease [unspecific]", "Lung Fibrosis",
    "Lung Injury [unspecific]", "Lung Neoplasms"
  )) %>%
  select(mir)


# Now sending those miRNAs to TargetScan
mirna_num <- seq(100, 800, 100)
mirna_target_num <- seq(2, 8, 1)
for (m in mirna_num[6:8]) {
  for (t in mirna_target_num[1:7]){
    mirna_genes <- mirna_calculator(ts_org = "Human", ts_version = "8.0", max_mir_targets = t, cancer_up = TRUE, cancer_type1 = "lung cancer", print_ts_targets = FALSE, max_mirnas = m, mirna_genes_mat_name = paste0("Outputs/increased_mirna_targets/mirna_genes_mat_", m, "_mirnas_", t, "_targets_up.rds"), mirna_ranking_name = paste0("Outputs/increased_mirna_targets/mirna_genes_", m, "_mirnas_", t, "_targets_up.rds"))
  }
}
```
