---
title: "SCLC Analysis"
author: "Andrew Willems & Tian Hong"
date: "2023-04-05"
output: html_document
---

```{r setup, include=FALSE}
# This sets the global chunk options for knitr to display the R code.
knitr::opts_chunk$set(echo = TRUE)

# This sets the root directory for knitr to the project directory.
knitr::opts_knit$set(root.dir = "~/Documents/Work/Phd_program/hong_lab/Projects/sclc/")
```

## SCLC Analysis

```{r load needed packages}
# Load necessary packages using the pacman package
# data.table: for efficient data manipulation
# devtools: for developing R packages
# ggplot2: for data visualization
# glmnet: for fitting generalized linear models via penalized maximum likelihood
# hoardr: for caching data to disk
# Matrix: for sparse matrix operations
# monocle3: for single-cell analysis and visualization
# phateR: for visualizing high-dimensional data
# reticulate: for running Python code in R
# Rmagic: for running R code in Jupyter notebooks
# styler: for consistent code formatting
# survminer: for survival analysis and visualization
# tidyverse: for data manipulation and visualization
# viridis: for creating color palettes for data visualization

pacman::p_load(
  data.table, devtools, ggplot2, glmnet, hoardr, Matrix, monocle3, phateR, 
  reticulate, Rmagic, styler, survminer, tidyverse, viridis
)
```

```{r miniconda setup for MAGIC}
# Loading conda environment 'r-reticulate' for MAGIC package
use_condaenv("/Users/andrewwillems/Library/r-miniconda-arm64/envs/r-reticulate")
```


```{r sourcing needed functions}
# This code sources several R scripts that contain functions used for analyzing miRNA expression data in lung cancer.
# The 'back_elim.R' script contains a function for performing backwards variable selection on the Cox model to identify the most significant miRNAs associated with survival.
# The 'cox_model.R' script contains a function for performing a Cox proportional hazards model on the data.
# The 'km_plotter.R' script contains a function for creating a Kaplan-Meier plot to visualize survival data.
# The 'lung_mirna_calculator.R' script contains functions for preprocessing and normalizing the miRNA expression data.
# The 'risk_score_calculator.R' script contains a function for calculating a risk score for each patient based on the results of the Cox model.
# This code also sources 'cell_dataset_builder.R', which contains functions for building a dataset from processed scRNA-seq expression data.

source("Code/back_elim.R")
source("Code/cell_dataset_builder.R")
source("Code/cox_model.R")
source("Code/km_plotter.R")
source("Code/lung_mirna_calculator.R")
source("Code/risk_score_calculator.R")
```


```{r reading data}
# Load gene expression data from a file and select relevant columns
scl_df <- read.csv(
  file = "Data/sclc_ucologne_2015/data_mrna_seq_rpkm.txt",
  sep = "\t"
)
scl_df <- subset(scl_df,
  select = c(Hugo_Symbol, sclc_ucologne_2015_S00022:sclc_ucologne_2015_S02397)
)

# Clean up column names by removing prefix
colnames(scl_df) <- sub("^sclc_ucologne_2015_S", "", colnames(scl_df))

# Load clinical data for patients and clean up column names and values
scl_patient <- read.csv(
  file = "Data/sclc_ucologne_2015/data_clinical_patient.txt",
  sep = "\t"
)
scl_patient <- scl_patient %>%
  slice(5:121) %>%
  rename(
    Patient.ID = X.Patient.Identifier, Overall.Survival.Status = Overall.Survival.Status,
    Overall.Survival.Months. = Overall.Survival..Months., Smoker = Smoker
  ) %>%
  select(Patient.ID, Sex, Overall.Survival.Status, Overall.Survival.Months., Smoker) %>%
  filter(Smoker != "Never") %>%
  filter(Smoker != "") %>%
  mutate(
    Smoker = as.factor(Smoker),
    Smoker = if_else(Smoker == "Current", true = 1, false = 0)
  ) %>%
  mutate(Overall.Survival.Status = ifelse(Overall.Survival.Status == "1:DECEASED", 1, 0)) %>%
  rename(
    vital.status = Overall.Survival.Status,
    time = Overall.Survival.Months.
  ) %>%
  mutate(Patient.ID = gsub(
    x = Patient.ID,
    pattern = "sclc_ucologne_2015_S",
    replacement = ""
  )) %>%
  mutate(vital.status_new = as.numeric(vital.status)) %>%
  mutate(time_new = as.numeric(time)) %>%
  select(-vital.status, -time) %>%
  rename(
    vital.status = vital.status_new,
    time = time_new
  ) %>%
  mutate(time = time * 30) %>%
  filter(time != 0) %>%
  rename_all(tolower)

# Extract the list of gene symbols
scl_genes <- scl_df$Hugo_Symbol

# Find the list of patients that have data for both gene expression and clinical data
common_patients <- intersect(scl_patient$patient.id, colnames(scl_df))

# Select only the common columns from gene expression data and transpose it
# so that each row corresponds to a patient
scl_common <- scl_df %>%
  select(all_of(common_patients)) %>%
  t() %>%
  as.data.frame()

# Assign the gene symbols as column names
colnames(scl_common) <- scl_genes

# Remove duplicated columns and add patient ID as a column
scl_common <- scl_common %>%
  select(-which(duplicated(names(.)))) %>%
  rownames_to_column(var = "patient.id")

# Join the gene expression data with the clinical data for each patient
scl_common <- left_join(scl_common, scl_patient[, c(
  "sex", "smoker",
  "vital.status", "time",
  "patient.id"
)],
by = "patient.id"
)

# Removing intermediate files to keep the environment clean
rm(scl_df, scl_patient, scl_genes, common_patients)
```


```{r single-cell data loading}
# Read in the gene expression data and filter out columns that start with "SCLC_P1" or "SCLC_P14"
sc_df <- fread("Data/OMIX002441-01.csv")
sc_df <- sc_df %>%
  select(-matches("^(SCLC_P1|SCLC_P14)")) %>%
  t()  # transpose the data frame to have cells as rows

# Set column names to the first row of the data frame
colnames(sc_df) <- sc_df["V1", ]
sc_df <- sc_df[-1, ]  # remove first row, which is now redundant
sc_df <- as.data.frame(sc_df)

# Read in the metadata and filter for tumor samples only
meta_data <- fread("Data/OMIX002441-02.csv")
meta_data <- meta_data %>%
  filter(NT == "tumor")

# Set row names to the first column, which contains the cell IDs
colnames(meta_data)[1] <- "id"

# Join the two dataframes on the "id" column
sc_df$id <- rownames(sc_df)
sc_df <- inner_join(sc_df, meta_data, by = "id")
rownames(sc_df) <- sc_df$id
sc_df <- sc_df %>% select(-id, -patient, -NT, -tissue, -cell_type, -nCount_RNA, -nFeature_RNA, -Pre_Map_Reads, -Aligned_Reads, -MappingRate)

# Keep only genes that are expressed in at least 10 cells
keep_cols <- colSums(sc_df > 0) > 10
sc_df <- sc_df[, keep_cols]

sc_rownames <- rownames(sc_df)

# Convert the data frame to integers
sc_df <- apply(sc_df, c(1, 2), as.integer) %>%
  as.data.frame()

rownames(sc_df) <- sc_rownames

# Plot the distribution of library sizes
ggplot() +
  geom_histogram(aes(x = rowSums(sc_df)), bins = 50) +
  geom_vline(xintercept = 1000, color = "red")

# Keep only cells with a library size between 1000 and 15000
keep_rows <- rowSums(sc_df) > 1000 & rowSums(sc_df) < 15000
sc_df <- sc_df[keep_rows, ]

# Normalize the library size and take the square root of the expression values
sc_df <- library.size.normalize(sc_df)
sc_df <- sqrt(sc_df)

# Removing intermediate files to keep the environment tidy
rm(keep_cols, keep_rows)
```

```{r single-cell denoising}
# Convert the sc_df dataframe to a matrix
sc_df_mat <- as.matrix(sc_df)

# Apply the MAGIC (Model-based Analysis of Genome-wide CRISPR/Cas9 Knockout) algorithm to denoise the scRNA-seq data.
# The 'genes' parameter specifies the set of genes to use for MAGIC imputation. Here, 'all_genes' is used to include all genes in the analysis.
# The 'seed' parameter specifies a random seed for reproducibility.
# The 'n.jobs' parameter specifies the number of cores to use for parallelization. Setting it to -1 indicates that all available cores should be used.
sc_denoised <- magic(sc_df_mat, genes = "all_genes", seed = 123, n.jobs = -1)
```


```{r generating pseudotime from single-cell data}
sc_denoised$result <- sc_denoised$result %>%
  t()

meta_data <- meta_data %>%
  filter(id %in% colnames(sc_denoised$result)) %>%
  column_to_rownames(var = "id")

cds_cell_meta <- meta_data
cds_gene_meta <- data.frame(gene_short_name = rownames(sc_denoised$result))
rownames(cds_gene_meta) <- cds_gene_meta$gene_short_name

cds_output <- cell_dataset_builder(
  vim_genes = cds_gene_meta$gene_short_name,
  cell_data = sc_denoised$result,
  cell_meta = cds_cell_meta,
  gene_meta = cds_gene_meta,
  my_root = "Y_7",
  point_size = 2.5,
  norm_flag = "log",
  my_cds_filename = "Data/cds_ouput.rds",
  my_monocle_graph = "Data/vim_pseudotime.svg",
  my_monocle_graph_genes = "Data/vim_genes.svg",
  my_pt_data_filename = "Data/vim_pseudotime_data.csv"
)

cds_output$`PT Graph`
cds_output$`Cell Progression Graph`
```




```{r mirna}
head(cox_df_sub)
mirna_ranking <- readRDS("Outputs/mirna/finished_metric/lung_mirnas_100_targets.rds")
gene_sizes <- seq(2, 20, 1)
cindices <- rep(0, length(gene_sizes))
all_coefs <- rep(0, length(gene_sizes))
active_coefs <- rep(0, length(gene_sizes))
counter <- 1
for (gs in gene_sizes) {
  cox_model <- cox_model_fitter(
    gene_num = gs,
    cox_predictors = names(mirna_ranking)[1:gs],
    cox_df = cox_df,
    n_folds = 10,
    my_seed = 1,
    my_alpha = 1,
    tumor_stage = FALSE,
    tumor_n = FALSE,
    tumor_m = FALSE,
    save_coefs = TRUE,
    calc_auc = FALSE,
    my_filename = "Outputs/cox_coefs/sclc_all_sexes_small_number_of_coefs_100_targets.csv"
  )

  cindices[counter] <- cox_model$CV$cvm[which.max(cox_model$CV$cvm)]
  all_coefs[counter] <- length(coef(cox_model$CV))
  active_coefs[counter] <-
    counter <- counter + 1
}

mirna_sclc_df <- data.frame(
  gene_size = gene_sizes,
  c_index = cindices,
  all_coefs = all_coefs
)

write.csv(mirna_sclc_df, file = "Outputs/cox_coefs/sclc_all_sexes_across_gene_size_small_number_of_coefs_100_targets.csv")
```
